File: v4/rust/iprange-livedb/src/validation/membership_table.rs
use std::mem;

use crate::cancellation::CancellationToken;
use crate::error::{Error, Result};

#[derive(Clone, Copy)]
pub(crate) struct Slot {
    pub(crate) id: u32,
    pub(crate) range_count: u64,
    pub(crate) stored_refcount: u64,
    word_count: u32,
    digest: [u8; 32],
    pub(crate) defined: bool,
    pub(crate) reverse_seen: bool,
}

impl Slot {
    const EMPTY: Self = Self {
        id: 0,
        range_count: 0,
        stored_refcount: 0,
        word_count: 0,
        digest: [0; 32],
        defined: false,
        reverse_seen: false,
    };
}

pub(crate) enum InsertResult {
    Inserted,
    Existing,
    Full,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum CountResult {
    Inserted,
    Existing,
    Full,
    Cancelled,
    Unavailable,
}

enum ProbeResult {
    Index(usize),
    Missing,
    Cancelled,
}

pub(crate) struct Table {
    slots: Vec<Slot>,
    mask: usize,
}

impl Table {
    pub(crate) fn new(entry_count: u64, available_bytes: u64) -> Result<Self> {
        if entry_count == 0 {
            return Ok(Self {
                slots: Vec::new(),
                mask: 0,
            });
        }
        let wanted = entry_count
            .checked_mul(2)
            .and_then(|value| value.checked_next_power_of_two())
            .ok_or(Error::ArithmeticOverflow(
                "validation membership table capacity",
            ))?;
        let capacity = usize::try_from(wanted)
            .map_err(|_| Error::BudgetExceeded("validation membership table"))?;
        let bytes =
            wanted
                .checked_mul(mem::size_of::<Slot>() as u64)
                .ok_or(Error::ArithmeticOverflow(
                    "validation membership table bytes",
                ))?;
        if bytes > available_bytes {
            return Err(Error::BudgetExceeded("validation membership table"));
        }
        let mut slots = Vec::new();
        slots
            .try_reserve_exact(capacity)
            .map_err(|_| Error::BudgetExceeded("validation membership table"))?;
        slots.resize(capacity, Slot::EMPTY);
        Ok(Self {
            slots,
            mask: capacity - 1,
        })
    }

    pub(crate) fn retained_bytes(&self) -> u64 {
        self.slots.len() as u64 * mem::size_of::<Slot>() as u64
    }

    pub(crate) fn count_range(&mut self, id: u32, cancellation: &CancellationToken) -> CountResult {
        let index = match self.probe(id, true, cancellation) {
            ProbeResult::Index(index) => index,
            ProbeResult::Missing => return CountResult::Full,
            ProbeResult::Cancelled => return CountResult::Cancelled,
        };
        let result = if self.slots[index].id == 0 {
            self.slots[index].id = id;
            CountResult::Inserted
        } else {
            CountResult::Existing
        };
        // A membership's count cannot exceed the total ranges addressable by
        // u32 pages and u16 page slots, so this cannot overflow u64.
        self.slots[index].range_count += 1;
        result
    }

    pub(crate) fn define(
        &mut self,
        id: u32,
        stored_refcount: u64,
        word_count: u32,
        digest: [u8; 32],
        cancellation: &CancellationToken,
    ) -> Result<InsertResult> {
        let Some(index) = self.find_or_empty(id, cancellation)? else {
            return Ok(InsertResult::Full);
        };
        if self.slots[index].id == 0 {
            self.slots[index].id = id;
        } else if self.slots[index].defined {
            return Ok(InsertResult::Existing);
        }
        let slot = &mut self.slots[index];
        slot.stored_refcount = stored_refcount;
        slot.word_count = word_count;
        slot.digest = digest;
        slot.defined = true;
        slot.reverse_seen = false;
        Ok(InsertResult::Inserted)
    }

    pub(crate) fn mark_reverse(
        &mut self,
        id: u32,
        word_count: u32,
        digest: [u8; 32],
        cancellation: &CancellationToken,
    ) -> Result<bool> {
        let Some(index) = self.find(id, cancellation)? else {
            return Ok(false);
        };
        let slot = &mut self.slots[index];
        if !slot.defined
            || slot.word_count != word_count
            || slot.digest != digest
            || slot.reverse_seen
        {
            return Ok(false);
        }
        slot.reverse_seen = true;
        Ok(true)
    }

    pub(crate) fn len(&self) -> usize {
        self.slots.len()
    }

    pub(crate) fn slot(&self, index: usize) -> Option<Slot> {
        let slot = *self.slots.get(index)?;
        if slot.id == 0 {
            return None;
        }
        Some(slot)
    }

    fn find(&self, id: u32, cancellation: &CancellationToken) -> Result<Option<usize>> {
        self.probe_result(id, false, cancellation)
    }

    fn find_or_empty(&self, id: u32, cancellation: &CancellationToken) -> Result<Option<usize>> {
        self.probe_result(id, true, cancellation)
    }

    fn probe_result(
        &self,
        id: u32,
        accept_empty: bool,
        cancellation: &CancellationToken,
    ) -> Result<Option<usize>> {
        match self.probe(id, accept_empty, cancellation) {
            ProbeResult::Index(index) => Ok(Some(index)),
            ProbeResult::Missing => Ok(None),
            ProbeResult::Cancelled => Err(Error::Cancelled),
        }
    }

    fn probe(&self, id: u32, accept_empty: bool, cancellation: &CancellationToken) -> ProbeResult {
        if self.slots.is_empty() || id == 0 {
            return ProbeResult::Missing;
        }
        let mut index = hash(id) & self.mask;
        for probe in 0..self.slots.len() {
            if probe != 0 && probe & 63 == 0 && cancellation.is_cancelled() {
                return ProbeResult::Cancelled;
            }
            let found = self.slots[index].id;
            if found == id || (accept_empty && found == 0) {
                return ProbeResult::Index(index);
            }
            if found == 0 {
                return ProbeResult::Missing;
            }
            index = (index + 1) & self.mask;
        }
        ProbeResult::Missing
    }
}

fn hash(id: u32) -> usize {
    (id.wrapping_mul(0x9e37_79b1)) as usize
}

#[cfg(test)]
mod tests {
    use std::sync::atomic::{AtomicUsize, Ordering};
    use std::sync::Arc;

    use super::*;

    #[test]
    fn long_probe_chain_has_bounded_cancellation_checkpoints() {
        let mut table = Table::new(64, u64::MAX).unwrap();
        let ready = CancellationToken::new();
        for index in 0..64 {
            assert_ne!(
                table.count_range(1 + index * 128, &ready),
                CountResult::Cancelled
            );
        }

        let polls = Arc::new(AtomicUsize::new(0));
        let observed = Arc::clone(&polls);
        let cancellation = CancellationToken::from_poll(Arc::new(move || {
            observed.fetch_add(1, Ordering::SeqCst);
            true
        }));
        assert!(matches!(
            table.count_range(1 + 64 * 128, &cancellation),
            CountResult::Cancelled
        ));
        assert_eq!(polls.load(Ordering::SeqCst), 1);
    }

    #[test]
    fn ordinary_probe_does_not_repeat_the_callers_checkpoint() {
        let mut table = Table::new(1, u64::MAX).unwrap();
        let polls = Arc::new(AtomicUsize::new(0));
        let observed = Arc::clone(&polls);
        let cancellation = CancellationToken::from_poll(Arc::new(move || {
            observed.fetch_add(1, Ordering::SeqCst);
            false
        }));

        assert!(matches!(
            table.count_range(1, &cancellation),
            CountResult::Inserted
        ));
        assert_eq!(polls.load(Ordering::SeqCst), 0);
    }
}

File: v4/go/internal/validation/membership_table.go
package validation

// Reverse-index table of one validation sweep (Rust
// validation/membership_table.rs): a bounded open-addressing probe
// table over the membership (or structure) ids, charged against the
// validation heap budget. The plan hash is the Rust u32 mix; probe
// chains checkpoint at the same 64-step cadence.

import (
	"unsafe"

	"github.com/firehol/iprange/v4/go/internal/format"
)

// Slot is one occupied table entry (Rust Slot).
type Slot struct {
	ID           uint32
	RangeCount   uint64
	StoredRefcnt uint64
	WordCount    uint32
	Digest       [32]byte
	Defined      bool
	ReverseSeen  bool
}

// InsertResult reports the outcome of one define (Rust InsertResult).
type InsertResult uint8

const (
	InsertInserted InsertResult = iota
	InsertExisting
	InsertFull
)

// CountResult reports the outcome of one range count (Rust
// CountResult).
type CountResult uint8

const (
	CountInserted CountResult = iota
	CountExisting
	CountFull
	CountCancelled
	CountUnavailable
)

// probeResult is the internal probe outcome (Rust ProbeResult).
type probeResult uint8

const (
	probeIndex probeResult = iota
	probeMissing
	probeCancelled
)

// Table is one bounded reverse-index probe table (Rust Table).
type Table struct {
	slots []Slot
	mask  uint
}

var emptySlot = Slot{}

// newTable builds the probe table with next-power-of-two capacity
// (Rust Table::new; ArithmeticOverflow and BudgetExceeded classes as
// exact).
func newTable(entryCount uint64, availableBytes uint64) (*Table, error) {
	if entryCount == 0 {
		return &Table{}, nil
	}
	wanted := entryCount * 2
	if wanted < entryCount {
		return nil, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation membership table capacity"}
	}
	// next power of two
	next := uint64(1)
	for next < wanted {
		next <<= 1
	}
	if next < wanted {
		return nil, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation membership table capacity"}
	}
	capacity := next
	if uint64(uint(capacity)) != capacity {
		return nil, &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "validation membership table"}
	}
	bytes := capacity * uint64(slotSize)
	if capacity != 0 && bytes/capacity != uint64(slotSize) {
		return nil, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation membership table bytes"}
	}
	if bytes > availableBytes {
		return nil, &format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "validation membership table"}
	}
	return &Table{slots: make([]Slot, capacity), mask: uint(capacity - 1)}, nil
}

// slotSize is the Go size of one Slot (Rust mem::size_of::<Slot>; the
// Go layout pads to the same 64 bytes as the Rust repr over the
// u32/u64 pair, the 32-byte digest, and the two flags).
var slotSize = int(unsafe.Sizeof(Slot{}))

// retainedBytes is the heap retained by the table (Rust
// Table::retained_bytes).
func (t *Table) retainedBytes() uint64 { return uint64(len(t.slots)) * uint64(slotSize) }

// countRange counts one range for id (Rust Table::count_range; the
// bounded probe can only report Full when the table is saturated).
func (t *Table) countRange(id uint32, check func() error) CountResult {
	index, outcome := t.probe(id, true, check)
	switch outcome {
	case probeIndex:
		if t.slots[index].ID == 0 {
			t.slots[index].ID = id
			t.slots[index].RangeCount++
			return CountInserted
		}
		t.slots[index].RangeCount++
		return CountExisting
	case probeMissing:
		return CountFull
	default:
		return CountCancelled
	}
}

// define records the stored facts of one dictionary entry (Rust
// Table::define).
func (t *Table) define(id uint32, storedRefcount uint64, wordCount uint32, digest [32]byte, check func() error) (InsertResult, error) {
	index, outcome := t.probeResult(id, true, check)
	if outcome == probeMissing {
		return InsertFull, nil
	}
	if outcome == probeCancelled {
		return 0, &format.Error{Code: format.CodeCancelled, Detail: "validation cancelled"}
	}
	if t.slots[index].ID == 0 {
		t.slots[index].ID = id
	} else if t.slots[index].Defined {
		return InsertExisting, nil
	}
	t.slots[index].StoredRefcnt = storedRefcount
	t.slots[index].WordCount = wordCount
	t.slots[index].Digest = digest
	t.slots[index].Defined = true
	t.slots[index].ReverseSeen = false
	return InsertInserted, nil
}

// markReverse records one reverse-index observation (Rust
// Table::mark_reverse).
func (t *Table) markReverse(id uint32, wordCount uint32, digest [32]byte, check func() error) (bool, error) {
	index, outcome := t.probeResult(id, false, check)
	if outcome != probeIndex {
		return false, nil
	}
	slot := &t.slots[index]
	if !slot.Defined || slot.WordCount != wordCount || slot.Digest != digest || slot.ReverseSeen {
		return false, nil
	}
	slot.ReverseSeen = true
	return true, nil
}

// markReverseDigest records one reverse-index observation without the
// word-count comparison (Rust Table::mark_reverse_digest).
func (t *Table) markReverseDigest(id uint32, digest [32]byte, check func() error) (bool, error) {
	index, outcome := t.probeResult(id, false, check)
	if outcome != probeIndex {
		return false, nil
	}
	slot := &t.slots[index]
	if !slot.Defined || slot.Digest != digest || slot.ReverseSeen {
		return false, nil
	}
	slot.ReverseSeen = true
	return true, nil
}

// len reports the capacity (Rust Table::len).
func (t *Table) len() int { return len(t.slots) }

// slot returns one occupied entry (Rust Table::slot).
func (t *Table) slot(index int) (Slot, bool) {
	if index < 0 || index >= len(t.slots) {
		return Slot{}, false
	}
	slot := t.slots[index]
	if slot.ID == 0 {
		return Slot{}, false
	}
	return slot, true
}

// probeResult maps the probe outcome to (index, ok) semantics (Rust
// find/find_or_empty).
func (t *Table) probeResult(id uint32, acceptEmpty bool, check func() error) (int, probeResult) {
	index, outcome := t.probe(id, acceptEmpty, check)
	if outcome == probeCancelled {
		return 0, probeCancelled
	}
	if outcome == probeMissing {
		return 0, probeMissing
	}
	return index, probeIndex
}

// probe is the bounded open-addressing probe (Rust Table::probe; an
// empty table or the zero id is always Missing; the probe-chain
// cancellation checkpoints every 64 steps).
func (t *Table) probe(id uint32, acceptEmpty bool, check func() error) (int, probeResult) {
	length := len(t.slots)
	if length == 0 || id == 0 {
		return 0, probeMissing
	}
	index := int(hash(id)) & int(t.mask)
	for step := 0; step < length; step++ {
		if step != 0 && step&63 == 0 && check != nil {
			if err := check(); err != nil {
				return 0, probeCancelled
			}
		}
		found := t.slots[index].ID
		if found == id || (acceptEmpty && found == 0) {
			return index, probeIndex
		}
		if found == 0 {
			return 0, probeMissing
		}
		index = (index + 1) & int(t.mask)
	}
	return 0, probeMissing
}

// hash is the plan u32 mixer (Rust hash: id * 0x9e3779b1).
func hash(id uint32) uint {
	return uint(id * 0x9e37_79b1)
}
